Project13

mainmeatball opened the pull request.

mainmeatball committed:
Add TypeScript support with rollup configuration

mainmeatball committed:
Add Project 13

[Grocery Cart Project](https://jsbeginners.com/grocery-cart-javascript-project/)

VagrantAI-c commented on Project 13/entity/card.ts:
```diff
@@ -0,0 +1,16 @@
+export class Card {
+    img: HTMLImageElement;
```
Здесь и далее - необязательно использовать строгие привязки к конкретному элементу, если не используются их специфические особенности. Такой подход позволит нивелировать изменчивую дом структуру: div может поменяться на p и тогда придется выискивать везде подобные вхождения.  `HTMLElement` достаточно в 95% сценариев.

VagrantAI-c commented on Project 13/entity/card.ts:
```diff
@@ -0,0 +1,16 @@
+    price: HTMLParagraphElement;
+    shopIcon: HTMLDivElement;
+
+    constructor(img: HTMLImageElement,
```
```typescript
class A { 
  a: string;
  constructor(a: string) { this.a = a; }
}
```
равноценно
```typescript
class A { constructor(public a: string) {} }
```

Это чисто тайпскриптовая фишка

VagrantAI-c commented on Project 13/entity/element-listener.ts:
```diff
@@ -0,0 +1,21 @@
+export namespace Project13 {
```
с приходом es6 модулей - неймспейсы стали не нужны

VagrantAI-c commented on Project 13/entity/element-listener.ts:
```diff
@@ -0,0 +1,21 @@
+            this.boundCallback = boundCallback;
+        }
+
+        public listen() {
```
```suggestion
        public listen(): void {
```
здесь и далее - функции которые ничего не возвращает можно помечать как void. Дело кодстайла делать так или нет: главное единообразие (также как например проставлять public кейворд или нет)

VagrantAI-c commented on Project 13/entity/element-listener.ts:
```diff
@@ -0,0 +1,21 @@
+export namespace Project13 {
+    export class ElementListener {
+        private field: HTMLElement;
```
чем филд провинился не быть ридонли?

VagrantAI-c commented on Project 13/entity/page.ts:
```diff
@@ -0,0 +1,42 @@
+
+    setCards(): void {
+        Shop.allVehicles
+            .map(vehicle => {
```
```suggestion
            .map((vehicle: Vehicle) => {
```
предпочитаю более вербозный вариант, поскольку из контекста не всегда понятно о каком типе идет речь, а также это позволит в случае чего при разработке подглянуть в структуру сущности. Опять же - делать так или нет - дело кодстайла и единообразия.

VagrantAI-c commented on Project 13/entity/page.ts:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    toggleShoppingList() {
+        this.shop.shoppingWindow.classList.toggle('move-x');
```
Чем давать прямой доступ к хтмл элементу другого класса, лучше инкапсулировать через функцию в `Shop`:
`this.shop.toggleShoppingList();`

VagrantAI-c commented on Project 13/entity/shop.ts:
```diff
@@ -0,0 +1,104 @@
+    private totals: HTMLParagraphElement[];
+    cartPreview: HTMLDivElement;
+    shoppingWindow: HTMLDivElement;
+    private shoppingWindowItemsContainer: HTMLDivElement;
```
```suggestion
    private readonly shoppingWindowItemsContainer = this.shoppingWindow.querySelector('#items');
```
Не стоит упускать возможности сделать ридонли филд

VagrantAI-c commented on Project 13/entity/shop.ts:
```diff
@@ -0,0 +1,104 @@
+        this.addShoppingWindowItem(item);
+    }
+
+    private incrementQuantity() {
```
Следует паблик функции размещать перед прайват функциями

VagrantAI-c commented on Project 13/entity/shop.ts:
```diff
@@ -0,0 +1,104 @@
+
+    public addVehicleToCart(item: Vehicle) {
+        this.incrementQuantity();
+        this.addTotal(+(item.price.replace(/[.$]/g, '')));
```
на такой случай полезно хранить в сущности цифровую версию цены и лейбл преобразовывать в формате хелпер функции `processPrice(vehicle: Vehicle): string; // 5000 -> '5.000$'`, либо рядышком хранить готовый лейбл преобразованный руками, чем парсить стринг в цену. Такой подход позволит стабилизировать иерархию сущностей: 
- цена это всегда число
- если надо такое число показать красиво, то делается функция-преобразователь
- если надо такое число показать иным образом во втором, третьем и др. местах, то делаются еще дополнительные функции-преобразователи

И никакие регекспы не понадобятся.

VagrantAI-c commented on Project 13/entity/shop.ts:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private incrementQuantity() {
+        this.quantity.textContent = '' + ++this.totalItems;
```
```suggestion
        this.quantity.textContent = `${++this.totalItems}`;
```
конвертация в строку из es5 формата в формат es6

VagrantAI-c commented on Project 13/entity/shopping-item.ts:
```diff
@@ -0,0 +1,50 @@
+    domElement: HTMLDivElement;
+    listeners: ElementListener[];
+
+    public static of(element: Element, shop: Shop): ShoppingItem {
```
Два варианта как писать подобные конструкции: выносить в хелпер функции/константы или делать статические функции/филды.

Плюсы статического подхода:
- декларативное описание: видно что к какой сущности относится

Плюсы хелпер функций/глобальных констант:
- самобытные структуры легко поддаются три шейкингу что имеет стратегическое преимущество в энтерпрайз разработке
- не усложняют исходную сущность, оставляя в ней только то, что ей необходимо в паблик апи (принцип разделения ответственности)
- если в хелпер функции используются несколько сущностей в аргументах/ретурн типах, то не надо гадать к какой сущности определить такую функцию, если бы мы делали ее статической

В силу данных причин я предпочитаю вариант с хелпер функциями. В твоем случае - особо без разницы.

mainmeatball replied:
Решил оставить как есть

VagrantAI-c commented on Project 13/index.css:
```diff
@@ -0,0 +1,211 @@
+	position: sticky;
+	top: 0;
+	left: 0;
+	z-index: 50;
```
Лучше не злоупотреблять z-index и выбирать более приближенные к реальности значения: 0 или 1. Иные значения должны систематизироваться и иметь пояснение почему было выбрано такое число. Почему здесь должно быть 50 а не 1?

VagrantAI-c commented on Project 13/index.css:
```diff
@@ -0,0 +1,211 @@
+	background-color: rgba(117, 117, 94, 0.95);
+	right: 0;
+	top: 96px;
+	z-index: 999;
```
Аналогичный вопрос про z-index здесь

VagrantAI-c commented on Project 13/index.css:
```diff
@@ -0,0 +1,211 @@
+}
+
+.fa-trash-o {
+	transition: color 0.3s linear;
```
Как правило ease и ее разновидности более привычно смотрится: инерция это естественное природное явление и ее аналог в веб-среде позволит пользователю быстрее освоиться: мы, люди, лучше понимаем интерфейс если видим знакомые нам паттерны из реального мира.

VagrantAI-c commented on Project 13/index.html:
```diff
@@ -0,0 +1,54 @@
+</head>
+<body onload="app.page.setCards()">
+	<div class="header">
+		<div id="cart-preview" class="cart-preview" onclick="app.page.toggleShoppingList()">
```
Если что-то должно кликаться, то этот элемент должен быть кнопкой

VagrantAI-c commented on Project 13/index.ts:
```diff
@@ -0,0 +1,18 @@
+    <HTMLTemplateElement>document.getElementById('cart-item-template')
+);
+
+const grid: Grid = new Grid(
```
```suggestion
const grid = new Grid(
```
можно не указывать тип: он вычислен тайпскриптом неявно при инициализации филда

VagrantAI-c reviewed (changes requested):
Неплохое начало в тайпскрипте, я думал будет больше моментов где можно было бы оптимизировать типы, но в целом все ок. С другой стороны особо сложных структур и преобразований еще не было, поэтому типы в духе
```typescript
type FilterConverterFn<T extends FilterName = FilterName> = (value: string[]) => VariableRequestFilter[T];
type FilterDictionary = Partial<{[key in FilterName]: FilterConverterFn<key>}>;
```
тебе к счастью вряд ли встретятся

mainmeatball committed:
Merge branch 'master' into project13

mainmeatball committed:
Project 13: fix resources images path

mainmeatball committed:
Project 13: remove redundant namespace

- Actually, it was added according to ElementListener class export error, that somehow eliminated itself and I can't reproduce it

mainmeatball committed:
Project 13: simplify constructors by removing class fields

mainmeatball committed:
Project 13: explicitly set return types

mainmeatball committed:
Project 13: set type in lambda function

mainmeatball committed:
Project 13: encapsulate toggle class method into shop class

mainmeatball committed:
Project 13: use readonly whenever you can

mainmeatball committed:
Project 13: move field initialization to field from constructor

mainmeatball committed:
Project 13: set public methods before private

mainmeatball committed:
Project 13: remove types when initializing variable

mainmeatball committed:
Project 13: remove circular dependencies

mainmeatball committed:
Project 13: use es6 number to string conversion feature

mainmeatball committed:
Project 13: extract price utils into external class

mainmeatball committed:
Project 13: remove unnecessary z-indexes

mainmeatball committed:
Project 13: use ease instead of linear animation

mainmeatball committed:
Project 13: use button instead of div as it is clickable

VagrantAI-c reviewed (approved):
Апрув, но архитектура кажется чрезмерно усложненной. Сложно сказать что-то точечно по ней: тут проще переписать, что я попробую сделать и покажу чуть позже как это можно организовать.
